homeassistant/loader.py

```diff
@@ -933,22 +933,185 @@ def all_dependencies(self) -> set[str]:
+    async def resolve_dependencies(self) -> set[str] | None:
         """Resolve all dependencies."""
-        if self._all_dependencies_resolved is not None:
-            return self._all_dependencies_resolved
+        if self._all_dependencies is not None:
+            if isinstance(self._all_dependencies, Exception):
+                return None
+            return self._all_dependencies
+
+        result = await self.resolve_multiple_dependencies(self.hass, (self,))
+        return result.get(self.domain)
+
+    @staticmethod
+    async def resolve_multiple_dependencies(
+        hass: HomeAssistant, integrations: Iterable[Integration]
+    ) -> dict[str, set[str]]:
+        """Resolve all dependencies for integrations.
+
+        Detects circular dependencies and missing integrations.
+        """
+        resolve_dependencies_tasks = {
+            itg.domain: create_eager_task(
+                itg._do_resolve_dependencies(cache_key="_all_dependencies"),  # noqa: SLF001
+                name=f"resolve dependencies {itg.domain}",
+                loop=hass.loop,
+            )
+            for itg in integrations
+        }
+
+        result = await asyncio.gather(*resolve_dependencies_tasks.values())
+
+        return {
+            domain: deps
+            for domain, deps in zip(resolve_dependencies_tasks, result, strict=True)
+            if deps is not None
+        }
+
+    @staticmethod
+    async def resolve_multiple_after_dependencies(
+        hass: HomeAssistant,
+        integrations: Iterable[Integration],
+        possible_after_dependencies: set[str] | None = None,
+        *,
+        ignore_exceptions: bool | None = None,
+    ) -> dict[str, set[str]]:
+        """Resolve all dependencies, including after_dependencies, for integrations.
+
+        Detects circular dependencies and missing integrations.
+        """
+        resolved: dict[str, set[str] | Exception] = {}
+
+        resolve_dependencies_tasks = {
+            itg.domain: create_eager_task(
+                itg._do_resolve_dependencies(  # noqa: SLF001
+                    cache=resolved,
+                    possible_after_dependencies=possible_after_dependencies,
+                    ignore_exceptions=ignore_exceptions,
+                ),
+                name=f"resolve after dependencies {itg.domain}",
+                loop=hass.loop,
+            )
+            for itg in integrations
+        }
+
+        result = await asyncio.gather(*resolve_dependencies_tasks.values())
+
+        return {
+            domain: deps
+            for domain, deps in zip(resolve_dependencies_tasks, result, strict=True)
+            if deps is not None
+        }
+
+    async def _do_resolve_dependencies(
+        self,
+        *,
+        cache: dict[str, set[str] | Exception] | None = None,
+        cache_key: str | None = None,
+        possible_after_dependencies: set[str] | None | UndefinedType = UNDEFINED,
+        ignore_exceptions: bool | None = None,
```

Comment:
Do we really need three different ways to handle exceptions? Can't we let the caller deal with the exception, for example let `resolve_multiple_dependencies` return `dict[str, set[str] | Exception]` instead of `dict[str, set[str]]`